src/paginationCalls/pagedApiCaller.ts

```diff
@@ -31,18 +31,111 @@
 
 import {APICaller, ApiCallerSettings} from '../apiCaller';
 import {
-  GaxCall,
   GRPCCall,
   NextPageRequestType,
   SimpleCallbackFunction,
   UnaryCall,
+  RawResponseType,
+  RequestType,
 } from '../apitypes';
 import {APICallback} from '../apitypes';
 import {OngoingCall, OngoingCallPromise} from '../call';
 import {CallOptions} from '../gax';
 import {GoogleError} from '../googleError';
 
 import {PageDescriptor} from './pageDescriptor';
+import {resolve} from 'dns';
+
+/**
+ * ResourceCollector class implements asynchronous logic of calling the API call that supports pagination,
+ * page by page, collecting all resources (up to `maxResults`) in the array.
+ *
+ * Usage:
+ *   const resourceCollector = new ResourceCollector(apiCall, maxResults); // -1 for unlimited
+ *   resourceCollector.processAllPages(request).then(resources => ...);
+ */
+class ResourceCollector {
+  apiCall: SimpleCallbackFunction;
+  resources: Array<{}>;
+  maxResults: number;
+  resolveCallback: (resources: Array<{}>) => void;
+  rejectCallback: (err: Error) => void;
+
+  constructor(apiCall: SimpleCallbackFunction, maxResults = -1) {
+    this.apiCall = apiCall;
+    this.resources = [];
+    this.maxResults = maxResults;
+    this.resolveCallback = (resources: Array<{}>) => {
+      throw new Error('Undefined callback');
```

Comment:
Yes, I'm just setting some default value here (I want the callback to be always defined) and then set it in the very beginning of `processAllPages`.  An alternative (possibly better) is to make it `?` (so that it could be undefined) and use as `callback!`.